ci/README: add github-script policy (#453985)
NixOS/nixpkgs: ci/README.md

Run [`update-pinned.sh`](./update-pinned.sh) to update it.

## GitHub Actions specific code

> Some of the code is specific to GitHub Actions.

Author: Right, changed GitHub Actions to just GitHub.

> Btw note that this also means that code which doesn't interact with GitHub doesn't fall under this policy (e.g. `ci/eval`, `nixpkgs-vet`), and indeed I think it should be the goal to make the code as GitHub-independent as possible.

Absolutely. The intent is to clearly split the github specific code from the forge agnostic code. Let's assume a hypothetical migration to a different forge: In this scenario, one should only need to replace the `ci/github-script` folder, because it is forge specific, but not any of the other things.

At the same time, I believe in both of "make code forge agnostic where possible" *and* "deeply integrate with the current forge". A clear split allows us to do that.

This code is currently spread out over multiple places and written in both Bash and JavaScript.
The goal is to eventually have all GitHub specific code in `ci/github-script` and written in JavaScript via `actions/github-script`.
A lot of code has already been migrated, but some Bash code still remains.

> New code must not be introduced as Bash, but as JavaScript instead.

Reviewer: I hope we are still ok with "trivial" (e.g. one-liner) steps being bash, e.g.:

- https://github.com/NixOS/nixpkgs/blob/127dea8a131f8706e2f0cf9997f9d0d2820a48c5/.github/workflows/backport.yml#L47

Or direct command invocation that is unrelated to github APIs, e.g.:
- https://github.com/NixOS/nixpkgs/blob/127dea8a131f8706e2f0cf9997f9d0d2820a48c5/.github/workflows/check.yml#L42
- https://github.com/NixOS/nixpkgs/blob/127dea8a131f8706e2f0cf9997f9d0d2820a48c5/.github/workflows/build.yml#L68
- https://github.com/NixOS/nixpkgs/blob/127dea8a131f8706e2f0cf9997f9d0d2820a48c5/.github/workflows/check.yml#L93

I'm not sure how to convey this nuance in the wording, but my interpretation of your current wording would be that the above examples are not permitted.

> New code must not be introduced as Bash, but as JavaScript instead.

Reviewer 2: ```suggestion
New CI features need to be introduced in JavaScript, not Bash.
```

Author: I think "features" catches cases of "significantly expanding the scope of an existing bash script with more features" vs "fixing a small bug in an existing bash script" better, where the former should be preceded by a migration of JavaScript, while the latter should be considered OK?
